### ISSUE TITLE
Matching Any with *sequence pattern produces unexpected reports

### ISSUE THREAD

**Issue by sorcio** [microsoft/pyright#5492]
I'm observing hard-to-explain behavior when I try to match `Any` with a sequence star pattern.

Easier to explain with a code example:

```python
def foo(x: Any) -> None:
    match x:
        case [*_sequence]:
            ...
        case _:  # error: Pattern will never be matched for subject type "Never"
            raise TypeError
```

I would expect the sequence pattern to only match sequences (as defined by [PEP 634](https://peps.python.org/pep-0634/#sequence-patterns)) and all sequences, and that's the behavior I see in Python 3.11.

Pyright instead seems to assume that the sequence pattern will match everything. But there seems to be more going on:

```python
def foo2(x: Any) -> None:
    match x:
        case [*_sequence]:
            return
    return 1  # error: ...
```

If the sequence pattern actually matched everything, the last line would be unreachable. Pyright instead considers code after the match statement to be reachable. This is correct, but seems to contradict the first observation.

Both the above cases behave the same replacing Any with object.

I'm adding one more snippet that might or might not be related:

```python
def foo3(x: Any) -> list[Any]:
    if isinstance(x, list):
        pass
    match x:
        case [*sequence]:  # error: Type of "sequence" is partially unknown
                           # Type of "sequence" is "list[Unknown] | list[Any]"
            return sequence  # error: Return type, "list[Unknown] | list[Any]", is partially unknown
```

Without the `isinstance()` check, the `sequence` variable is taken to have type `list[Any]`. I'm only adding this because of the confusion produced by this report in this context, and because I thought there might be some special-casing with the combination of Any and sequence patterns. Replacing `x: Any` with `x: object` doesn't trigger the same error.

---
* Observed both with latest Pylance and pyright 1.1.317
* MyPy seems to confirm my expectations, so reporting this as a bug

**Comment by erictraut** [comment]
Thanks for the bug report. Yes, pyright is incorrectly handling type narrowing in the negative (fall-through) case for sequence patterns when the subject type is `Any`. I've added a fix, and this will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.318, which I just published. It will also be included in a future release of pylance.
